sql: support ON CONFLICT ... RETURNING

couchand opened the pull request:
This is a rough first pass at adding support for a `RETURNING` clause in `UPSERT` and `INSERT ... ON CONFLICT` statements.  Not ready to merge, putting this up for early feedback.  I'm playing fast and loose about allocations, but based on the surrounding code I have a feeling we'll want to be more thoughtful about them.  There are a number of TODOs, some of which should be taken care of here and some of which are just reminders of interface inconsistencies that could be cleaned up.

The general approach here is to have the upsert node run to completion, keeping the batch of all result rows, then doling them out one at a time on calls to `Values`.

I chatted with @knz about this and I agree with him that this approach leaves something to be desired: it would be much cleaner to get rid of the `returningHelper` and make edit nodes act properly as data sources, adding a render node after them if they have a `RETURNING` clause.  However, this change was much easier for me to wrap my head around, and if we want to get support for this feature in for 1.1 this might be the way to go as it does seem to be smaller.  In any case I'll go ahead with looking at refactoring away the returning helpers to get a sense for what that will take.

(closes #6637)

cc: @cockroachdb/sql-execution

couchand committed:
sql: support ON CONFLICT ... RETURNING

This adds support for the combination of ON CONFLICT and RETURNING on
INSERT statements (as well as RETURNING on the UPSERT shorthand).  We
drain the data source and upsert all rows on the first call to Next,
collecting them in the insert node and doling them out one at a time on
calls to Values.

Closes #6637.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/17149)
<!-- Reviewable:end -->

knz commented:
1) ok for this "naive" approach, this moves us forward in terms of functionality. However you need to push a little bit to get the memory accounting right; we don't want to start accumulating data in memory without any limitation. Some suggestions below in comments for this.

2) INSERT ON CONFLICT has this notion of an `excluded` pseudo-table which contains the values "before the insert". 
Have you checked that RETURNING expressions correctly use the values from the *final* row as opposed to the values of the row before the conflict is resolved? (Same question for UPSERT really)

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/delete.go, line 156 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXBa0moB_lKJkpu2Jr:-KpXBa0moB_lKJkpu2Js:bibvvlh) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/delete.go#L156)):*
> ```Go
> 		if err == nil {
> 			// We're done. Finish the batch.
> 			// TODO(couchand): Use the first result?
> ```

Please extend the comment and explain what the phrase "first result" designates.

---

*[pkg/sql/insert.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXDf4tDgjYMS_7N6bt:-KpXDf4tDgjYMS_7N6bu:b84b58n) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L60)):*
> ```Go
> 	tw                    tableWriter
> 
> 	isUpsertReturning bool
> ```

I would simply name this `hasResultValues`

---

*[pkg/sql/insert.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXDY2U0eAkv7O9PWHS:-KpXDY2VpX1yAnovfjLx:bdc5ad) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L70)):*
> ```Go
> 
> 		doneUpserting bool
> 		rowsUpserted  []parser.Datums
> ```

you should use a `RowContainer` here to ensure proper memory accounting.

---

*[pkg/sql/insert.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXE5ULse2w14nsIT2r:-KpXE5UMfssZ2s9kd8NU:b-ot04bp) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L317)):*
> ```Go
> 		}
> 	}
> 	return len(n.run.rowsUpserted) > 0, nil
> ```

Please add a comment to explain the return value, or a comment above on the method to explain what it does.

---

*[pkg/sql/insert.go, line 360 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXEQK84Ya_4Qu5cUQC:-KpXEQK9Zcp5PsrVe4MJ:b-2h0syt) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L360)):*
> ```Go
> 	}
> 
> 	if !n.isUpsertReturning {
> ```

Please explain here why a result row is being prepared ("cooked") despite the fact "isUpsertReturning" is false.

---

*[pkg/sql/insert.go, line 545 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXEWAYEgRWGuFKQZla:-KpXEWAYEgRWGuFKQZlb:b19suzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L545)):*
> ```Go
> 
> func (n *insertNode) Values() parser.Datums {
> 	if n.isUpsertReturning {
> ```

It's more idiomatic in Go to have the short code path inside the conditional and the long path out, e.g.
```go
if !n.isUpsertReturning {
  return n.run.resultRow}
}
row := ...
```

---

*[pkg/sql/tablewriter.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXEned7G180Fewxf6I:-KpXEneeuRReI6RIjFyU:b-saaf52) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/tablewriter.go#L76)):*
> ```Go
> 	// should be logged to the context. See the comment above for why
> 	// this a separate parameter as opposed to a Value field on the context.
> 	finalize(ctx context.Context, traceKV bool) ([]parser.Datums, error)
> ```

1) return a `RowContainer` here.
2) Make the function take a boolean argument that indicates whether to collect the result rows or not (so that this cost is not paid when the rows are not needed).

---

*[pkg/sql/tablewriter.go, line 564 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXF2pFGV8ezXUGb0M6:-KpXF2pFGV8ezXUGb0M7:b23zsxz) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/tablewriter.go#L564)):*
> ```Go
> 			return nil, err
> 		}
> 		// TODO(couchand): Test this case.
> ```

yes please :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-:-KpXF7kw6RkvP2FmBMk7:b-objg21)*
<!-- Sent from Reviewable.io -->

couchand commented:
I believe I've taken care of the memory accounting now, and I've added a test case that the row after conflict resolution is the one that's returned.

I've addressed all the feedback below, and I believe this is now ready to merge.

---

Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/sql/delete.go, line 156 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXBa0moB_lKJkpu2Jr:-Kq4Eo1D3qZHixAqTOlw:bgqq1i) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/delete.go#L156)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please extend the comment and explain what the phrase "first result" designates.
</blockquote></details>

I've removed the comment as I don't think it's valid.

---

*[pkg/sql/insert.go, line 60 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXDf4tDgjYMS_7N6bt:-Kq4EqugoMyNDwKJ9pfG:b-ays7d4) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L60)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I would simply name this `hasResultValues`
</blockquote></details>

This isn't exactly the same as simply having a `RETURNING` clause, because an `INSERT` with a `RETURNING` but without an `ON CONFLICT` is handled separately.  I do think it would be nice to unify them, but that would be more complicated than I would like to handle in this PR.

---

*[pkg/sql/insert.go, line 70 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXDY2U0eAkv7O9PWHS:-Kq4FD0um4JYGE_Rfuxf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L70)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

you should use a `RowContainer` here to ensure proper memory accounting.
</blockquote></details>

Done.

---

*[pkg/sql/insert.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXE5ULse2w14nsIT2r:-Kq4F8J5SoRCYKPyR9AH:b-46w5ys) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L317)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please add a comment to explain the return value, or a comment above on the method to explain what it does.
</blockquote></details>

I've added an explanatory constant.

---

*[pkg/sql/insert.go, line 360 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXEQK84Ya_4Qu5cUQC:-Kq4FFXkDkA0jV8Z2mTh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L360)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please explain here why a result row is being prepared ("cooked") despite the fact "isUpsertReturning" is false.
</blockquote></details>

Done.

---

*[pkg/sql/insert.go, line 545 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXEWAYEgRWGuFKQZla:-Kq4FGoe9rnFImVWDL44:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L545)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

It's more idiomatic in Go to have the short code path inside the conditional and the long path out, e.g.
```go
if !n.isUpsertReturning {
  return n.run.resultRow}
}
row := ...
```
</blockquote></details>

Done.

---

*[pkg/sql/tablewriter.go, line 76 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXEned7G180Fewxf6I:-Kq4FHgiSBqLo5GMxzUJ:b-e0cur4) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/tablewriter.go#L76)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) return a `RowContainer` here.
2) Make the function take a boolean argument that indicates whether to collect the result rows or not (so that this cost is not paid when the rows are not needed).
</blockquote></details>

Done, though for (2) it's as a field on the `tableUpserter` struct, since it needs to be accessible in more than just `finalize`.

---

*[pkg/sql/tablewriter.go, line 564 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXF2pFGV8ezXUGb0M6:-Kq4FT-wMXDP_cm9kswm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/tablewriter.go#L564)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

yes please :)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-:-Kq4FUBGDV0mRxO5RoaC:bonizkf)*
<!-- Sent from Reviewable.io -->

couchand commented:
Can I get a final review from somebody on @cockroachdb/sql-execution please?

jordanlewis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod a couple of small suggestions if @knz's memory accounting concerns are resolved!

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[pkg/sql/insert.go, line 348 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KqOM46VsueNaIb8xnxV:-KqOM46VsueNaIb8xnxW:b-5bf85w) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f018249a67894887902174dc38887308db552ea/pkg/sql/insert.go#L348)):*
> ```Go
> 					params.p.session.TxnState.makeBoundAccount(),
> 					sqlbase.ColTypeInfoFromResCols(n.rh.columns),
> 					0,
> ```

Can we preallocate the row container by passing rows.Len() instead of 0 here?

---

*[pkg/sql/tablewriter.go, line 387 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KqOMXlfaSD8r-vuMzGu:-KqOMXlfaSD8r-vuMzGv:bcfl130) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f018249a67894887902174dc38887308db552ea/pkg/sql/tablewriter.go#L387)):*
> ```Go
> 		// In some cases (e.g. `INSERT INTO t (a) ...`) rowVals does not contain all the table
> 		// columns. We need to pass values for all table columns to rh, in the correct order; we
> 		// will use rowTemplate for this. We also need a table that maps row indices to rowTemplate indices
> ```

If you're already going to rebase this, could you please wrap the comments at 80 chars?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-:-KqON8PeBDBv-eJW1tCq:b9iymb9)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 5 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/sql/insert.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXE5ULse2w14nsIT2r:-KqOeRsh9vPNqP_JTuZC:b-6h8wqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L317)):*
<details><summary><i>Previously, couchand (Andrew Couch) wrote…</i></summary><blockquote>

I've added an explanatory constant.
</blockquote></details>

I don't see the constant, but I see a comment, I think you're fine :)

---

*[pkg/sql/insert.go, line 296 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KqOeGMF1tFWKCOHITLc:-KqOeGMF1tFWKCOHITLd:b-fd6xcg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7784ebccf31d15306b36239a7729dd09bc259e7f/pkg/sql/insert.go#L296)):*
> ```Go
> 	n.run.rows.Close(ctx)
> 	if n.run.rowsUpserted != nil {
> 		n.run.rowsUpserted.Close(ctx)
> ```

Strongly suggested:
```go
n.run.rows = nil
n.run.rowsUpserted = nil
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-:-KqOf5luWQKlG2BuI5Zf:bnfp4nl)*
<!-- Sent from Reviewable.io -->

couchand commented:
Review status: 3 of 5 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/insert.go, line 317 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KpXE5ULse2w14nsIT2r:-KqP8WHgVJ1XUqb9mh5X:btojbfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fc9dcbbf2883189375ada5e42cfd2832ab515/pkg/sql/insert.go#L317)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I don't see the constant, but I see a comment, I think you're fine :)
</blockquote></details>

Great.

---

*[pkg/sql/insert.go, line 348 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KqOM46VsueNaIb8xnxV:-KqP8XLCx7byM7vGGkYY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f018249a67894887902174dc38887308db552ea/pkg/sql/insert.go#L348)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can we preallocate the row container by passing rows.Len() instead of 0 here?
</blockquote></details>

Done.

---

*[pkg/sql/insert.go, line 296 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KqOeGMF1tFWKCOHITLc:-KqP8VCy0cHwIwSDbldO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7784ebccf31d15306b36239a7729dd09bc259e7f/pkg/sql/insert.go#L296)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Strongly suggested:
```go
n.run.rows = nil
n.run.rowsUpserted = nil
```
</blockquote></details>

Done.

---

*[pkg/sql/tablewriter.go, line 387 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149#-KqOMXlfaSD8r-vuMzGu:-KqP8YKLo6SrPJswP-LQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0f018249a67894887902174dc38887308db552ea/pkg/sql/tablewriter.go#L387)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

If you're already going to rebase this, could you please wrap the comments at 80 chars?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17149)*
<!-- Sent from Reviewable.io -->
